helix-core/src/case_conversion.rs

```diff
@@ -2,68 +2,328 @@ use crate::Tendril;
     }
 }
 
-pub fn to_lower_case_with(text: impl Iterator<Item = char>, buf: &mut Tendril) {
-    for c in text {
-        for c in c.to_lowercase() {
-            buf.push(c)
+pub fn into_alternate_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, |c| {
+        if c.is_uppercase() {
+            c.to_ascii_lowercase()
+        } else if c.is_lowercase() {
+            c.to_ascii_uppercase()
+        } else {
+            *c
         }
-    }
+    });
 }
 
-pub fn to_camel_case(text: impl Iterator<Item = char>) -> Tendril {
+pub fn into_upper_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, char::to_ascii_uppercase);
+}
+
+pub fn into_lower_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, char::to_ascii_lowercase);
+}
+
+pub fn into_kebab_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    separator_case_conversion(text, buf, '-');
+}
+
+pub fn into_snake_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    separator_case_conversion(text, buf, '_');
+}
+
+pub fn into_title_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    complex_case_conversion(text, buf, true, Some(' '));
+}
+
+pub fn into_camel_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    complex_case_conversion(text, buf, false, None);
+}
+
+pub fn into_pascal_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    complex_case_conversion(text, buf, true, None);
+}
+
+fn to_case<I>(text: I, to_case_with: fn(I, &mut Tendril)) -> Tendril
+where
+    I: Iterator<Item = char>,
+{
     let mut res = Tendril::new();
-    to_camel_case_with(text, &mut res);
+    to_case_with(text, &mut res);
     res
 }
-pub fn to_camel_case_with(mut text: impl Iterator<Item = char>, buf: &mut Tendril) {
-    for c in &mut text {
-        if c.is_alphanumeric() {
-            buf.extend(c.to_lowercase())
+
+pub fn to_camel_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_camel_case)
+}
+
+pub fn to_lower_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_lower_case)
+}
+
+pub fn to_upper_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_upper_case)
+}
+
+pub fn to_pascal_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_pascal_case)
+}
+
+pub fn to_alternate_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_alternate_case)
+}
+
+pub fn to_title_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_title_case)
+}
+
+pub fn to_kebab_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_kebab_case)
+}
+
+pub fn to_snake_case(text: impl Iterator<Item = char>) -> Tendril {
+    to_case(text, into_snake_case)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_camel_case_conversion() {
+        let tests = [
```

Comment:
Maybe, if there are others in Helix already, we could use parameterised tests? I know it requires some extra dependency, so if there are no other such tests, just resolve this comment.

Reply:
doesn't seem to be the case